Python improvements

ironfroggy opened the pull request.

Commit:
Python 2, weird manage.py placement, proper argument parsing, configurable Babel setup

Commit:
Document new setup command

dpoirier commented on README.md:
````diff
@@ -10,15 +10,36 @@ Jazz Hands is a automatic build and dev-run tool for Django projects with modern
+You can install Jazzhands locally in the virtual environment of a Django project.
+
+```
+pip install -e git@github.com:caktus/jazzhands.git
````
Shouldn't need ``-e`` unless someone wants to develop on jazzhands itself.

dpoirier replied:
That doesn't seem to be a valid argument to pip.  This works better:

    pip install 'git+https://github.com/caktus/jazzhands@python-improvements'

dpoirier commented on README.md:
````diff
@@ -10,15 +10,36 @@ Jazz Hands is a automatic build and dev-run tool for Django projects with modern
+jazzhands setup -p es2017
+```
+
+Multiple presets are allowed. For a React project you would add the `react` preset as well.
````
s/are allowed/are allowed by repeating the `-p` option/

(otherwise I would have guessed passing a comma-separated list of presets to a single `-p`)

dpoirier commented on README.md:
```diff
@@ -10,15 +10,36 @@ Jazz Hands is a automatic build and dev-run tool for Django projects with modern
 you can install prerequisite NPM packages for ECMAScript compilation with Jazzhand's `setup`
 command.
 
+The setup command needs to know what Babel presets and transform plugins you want. As a basic recommendation you should include the `es2017` preset, which will include all the ES2017 transformations necessary. The setup will add the necessary entries to your `package.json` file as well as generate a `.babelrc` configuration file for you.
```
How would I find out what the allowed Babel presets and transform plugins are?  (Maybe link to babel doc or something?)

ironfroggy replied:
Moved some of this around, along with making the ES2017 preset a default if you omit `-p`, and added the doc link suggested.

dpoirier commented on jazzhands/__main__.py:
```diff
@@ -184,25 +209,57 @@ def main(argv=sys.argv):
+    elif args.which == 'setup':
         # Do one time setup stuff for the project
-        args = [
+        # Currently defaults to Babel's ES2017
```
Is this comment accurate? I don't see es2017 in the list below.

ironfroggy replied:
fixed

dpoirier commented on jazzhands/__main__.py:
```diff
@@ -259,9 +321,9 @@ def main(argv=sys.argv):
-        if root.endswith('static/js'):
-            print('JS Destination', root)
-            js_dir = root
+        # if root.endswith('static/js'):
```
Any need to keep these commented lines?

dpoirier commented on jazzhands/__main__.py:
```diff
@@ -184,25 +209,57 @@ def main(argv=sys.argv):
+    parsers['run'].add_argument('--auto-npm', action='store_true')
+
+    args = parser.parse_args()
+
```
If I install this branch in voters-ingestor, and run jazzhands without arguments, it fails with a stack trace:

    $ jazzhands 
    Traceback (most recent call last):
      File "/home/dpoirier/.virtualenvs/voters-ingestor/bin/jazzhands", line 11, in <module>
        load_entry_point('jazzhands==0.1', 'console_scripts', 'jazzhands')()
      File "/home/dpoirier/.virtualenvs/voters-ingestor/lib/python3.6/site-packages/jazzhands/__main__.py", line 344, in main
        if index_files['js'] and js_dir:
    KeyError: 'js'

dpoirier replied:
I ran `jazzhands setup` apparently successfully in voters-ingestor, then `jazzhands collect` failed at the same place.

dpoirier replied:
I installed it in typecoach, cded to the webapp directory, and it ran without error without arguments, and did setup okay. Then `jazzhands collect` printed "Could not locate your project's main package. Jazzhands tries to locate this relative to a `settings` package or `settings.py` module."  There's a settings.py file in that directory.

ironfroggy replied:
It is run from the top of the repository and will locate everything. It has to be at the top to find everything relative to the project's root folder.

ironfroggy replied:
I also don't consider the failure on voters-ingestor to be a problem, because that project has no frontend to build so Jazzhands is not applicable.

dpoirier replied:
In typecoach, the whole Django app is under webapp, including manage.py and settings. I don't see why I shouldn't expect jazzhands to be able to find everything since it's all there.

ironfroggy replied:
The Django project is all under `webapp` but Jazzhands is concerned with more than just the Django project. The `.babelrc` configuration and `node_modules` needed for the frontend building, which is its primary purpose, need to be at the repository level where those things live.

dpoirier replied:
In that project, ``node_modules`` actually is under ``webapp``.  

In any case, it seems perfectly reasonable for a Django project to have a settings.py in its top-level directory. Is there some reason to disallow that?

dpoirier commented on jazzhands/__main__.py:
```diff
@@ -184,25 +209,57 @@ def main(argv=sys.argv):
+    parser = argparse.ArgumentParser()
+    parser.set_defaults(which=None)
+    # parser.add_argument('command', type=str, action='store')
+    subparsers = parser.add_subparsers()
```
Fancy!

dpoirier commented on jazzhands/__main__.py:
```diff
@@ -184,25 +209,57 @@ def main(argv=sys.argv):
+
+    add_command('setup')
+    parsers['setup'].add_argument('-p', '--preset', action='append', type=str, default=[])
+    parsers['setup'].add_argument('-t', '--transform', action='append', type=str, default=[])
```
I never knew you could do this with argparse.

ironfroggy replied:
Neither did I!

dpoirier commented on jazzhands/__main__.py:
```diff
@@ -184,25 +209,57 @@ def main(argv=sys.argv):
+        for transform in args.transform:
+            cargs.append("babel-plugin-transform-" + transform)
+
+        babelrc = {
```
Should we check whether a .babelrc file already exists and not overwrite it?  Maybe just print out what we would have done and prompt the user to update their .babelrc manually if necessary?

Commit:
Fixing numerous code review requests:

- Documenting pip install without -e
- Adding default Babel preset and updating documentation to match
- Don't overwrite .babelrc that already exists
- Remove some commented code and update comments about destinations

Commit:
Add full instructions to manual babel setup.

dpoirier commented on jazzhands/__main__.py:
```diff
@@ -193,6 +193,17 @@ def manage_py(args, background=False):
         os.chdir(cwd)
 
 
+def comma_and(args, quote=None):
```
Could we add a docstring to this method?

dpoirier reviewed (changes requested):
Docstring for comma_and method.

Commit:
Add checks to make sure jazzhands is run in the right context.

Commit:
Exit early if package.json check fails when trying to setup.

Commit:
Exit early if package.json check fails when trying to setup. (2)

dpoirier commented on jazzhands/__main__.py:
```diff
@@ -150,14 +159,15 @@ def build_less(dirs):
-    args.extend("-t [ babelify --presets [ react es2015 ] ]".split())
+    args.extend("-t [ babelify ]".split())
     args.extend([index_files['js'], '-o', os.path.join(js_dir, 'bundle.js')])
+    print(args)
```
debug?

dpoirier reviewed (changes requested):
Please use DJANGO_SETTINGS_MODULE to locate the settings, at least if it's set.  Django doesn't assume the settings file is anywhere in particular, and neither should jazzhands if it's going to be useful for arbitrary Django projects with frontends.

Commit:
Remove debug print
